### DISCUSSION TITLE
immich import existing library

### DISCUSSION THREAD

**Discussion by kaanvur** [alexbelgium/hassio-addons#1467] [Q&A]
immich was previously installed on docker. But I did a clean installation and moved it to the home assistant.
I moved all my files. I showed the way on the installed immich.
My existing images do not appear, but when I upload a new file, it works without any problems.
How can I import my existing gallery?

**Comment by alexbelgium** [comment]
Hi, i think immich documentation could help you better than me... You can access immich through command line with the portainer add-on, BTW. But I can help if there is an action to do from the add-on side

**Reply by kaanvur** [reply]
I saw that they only suggested adding it as an external library. But no path I enter is accepted by the external library.
Can you help me if you know a way?

**Reply by alexbelgium** [reply]
Do you know if the new files appear at the same place as your old librairies? What path are you using and what is the add-on startup log?

**Reply by kaanvur** [reply]
I have an external hard disk. I mounted it using samba nas. When installing immich, I used /media/WD/Gallery as the data location. When I upload a file from immich. It is uploaded to the /media/WD/Gallery/upload folder. My old photos are in /media/WD/Gallery/library/admin

**Reply by alexbelgium** [reply]
Immich uses 1000:1000 as permissions ; did you gave the same permissions to that folder? I don't remember if immich can run with PUID 0 PGID 0 you could try that

Or can you reimport your whole library? As anyway it still needs to be reindexed ?

**Reply by alexbelgium** [reply]
According to copilot there is several options : https://copilot.microsoft.com/sl/hvpEE3vCbyC

**Comment by ProxiBlue** [comment]
Hello, I have worked out a way to migrate existing immich to the add-on.

Requires a bit of technical skills, but I am sure people using HAOS + immich would have what is needed :)

See this gist:  https://gist.github.com/ProxiBlue/33f8d50266b845eba3a5c4a4b3eaac16

Process better explained:

Requires install of the terminal add-on and ability/means to copy / mount / place your migrated immich files where settings of immich designates the data storage is.

![image](https://github.com/user-attachments/assets/51e635bd-2ea9-4e92-9912-70f367b6c4cf)

In my case, I mounted an external USB, using the udev method, directly onto data folder. 

This assumes you started with a brand new add-on install, and all previous add-on data is deleted.

1. Install the immich postgreSQL add-on. 
2. After install, before you start it, edit settings to be as below.

![2024-12-31_02-16](https://github.com/user-attachments/assets/3feb640e-0a33-484f-a653-aa55231133a8)

3.  start the database service. 
wait 10s. 
stop. start again. 
I found this works. just that one initial start seems to cause connection issues.

4. using the terminal add-on, edit the autoscript and place the contents of this file in it.
![image](https://github.com/user-attachments/assets/2aa7172f-7b9f-4e35-bbe2-e926de7c7dd2)

```
vi /homeassistant/addons_autoscripts/immich.sh 
```
and then paste the contents (or use whatever editor you are comfortable with)


5. make sure it can run: ```chmod +x /homeassistant/addons_autoscripts/immich.sh```
![image](https://github.com/user-attachments/assets/2c84b826-e0d8-4861-b5fa-0275e39bb059)

6. Install immich add-on, or just start it if already installed.

my immich setup. make sure your username is correct as well as host. I just use the HOST machine IP.

![image](https://github.com/user-attachments/assets/ac6b6f0e-0218-4382-8fb0-38ec269d0bd7)


if all goes well, you should see the import happen in the immich start logs. Give it some time

![image](https://github.com/user-attachments/assets/6f751fde-aab8-4d1e-bb30-1b7d8de4e688)

**Reply by alexbelgium** [reply]
Thanks very much for sharing!

**Comment by ProxiBlue** [comment]
## A path issue I ran into. (see next comment it is fixed)

After import I could see all the data was there, but, no thumbnails or images showed. only placeholders

![image](https://github.com/user-attachments/assets/3122650e-f90e-4388-a638-ed29cffb71c6)

a bit of investigation, shows that immich is doing a really odd thing. 

They are placing the defined base path (upload_location) and saving that as part of the path to assets.
This seems a bit odd, as it will hamper things like migrations.
( I need to check if this was a raised issue ;/ )

immich add-on as a base install, with one test image uplaoded.

![image](https://github.com/user-attachments/assets/c7517e41-9992-44ee-857f-afe1a4174252)

notice the full path

and then, after import, my old db assets

![image](https://github.com/user-attachments/assets/4bebc282-8145-44f1-b503-5c5bba54dd30)

There, notice /upload/ in the path.

So, need to transpose this to the new path as used with HAOS add-on

To fix this, you need to connect to the db and run some SQL statements to migrate the paths. That is just but one way, which is what I followed. 

I just used my IDE and connected direct to the postgresql db, and ran the queries needed. You couldalso load the portainer add-on, access the immich container shell, and use psql on the cli. or, just use any pgsql client.

so, what you need is some sql to replace the paths. 
this will do, but you obviously need to adjust it to your paths

```
UPDATE assets
SET "originalPath" = REPLACE("originalPath", 'upload/', '')
WHERE "originalPath" LIKE '%upload/%';
```

as an example, I need to be careful, as I have paths like:  'upload/upload/' so that will break. I need to deal with them first.

```
UPDATE assets
SET "originalPath" = REPLACE("originalPath", 'upload/upload/', '/share/immich/PLACEHOLDER/')
WHERE "originalPath" LIKE 'upload/upload/%';
```

moved them out the way for now.

2nd query to move the bulk.

```
UPDATE assets
SET "originalPath" = REPLACE("originalPath", 'upload/', '/share/immich/')
WHERE "originalPath" LIKE 'upload/%';
```

and then move the placeholder back (as that is a valid upload/ folder

```
UPDATE assets
SET "originalPath" = REPLACE("originalPath", 'PLACEHOLDER', 'upload')
WHERE "originalPath" LIKE '%PLACEHOLDER%';
```

![image](https://github.com/user-attachments/assets/62c8206f-3723-483d-b256-9f2e0f4bfc18)

BUT, that did not work. I am not sure why not. strill. not a bad thingt o fix. Feels like a cache issue
so, I did an additional step.

I worked out (by a bit of trial and error!) that immich wants this path: 

```
/app/immich/server
```

don't ask me why.

so, I just used portainer shell, and placed  a symlink here.

```
ln -s /share/immich ./upload
```

and then it worked.

I suspect there is a cache involved, or possible another table I need to adjust to fix the paths.

![image](https://github.com/user-attachments/assets/0065c7c8-777b-4aff-88d5-9726ea3795ff)

Happy days !

**Comment by ProxiBlue** [comment]
Even more happy days.

The problem is solved by making the correct changes to the database.

```
UPDATE asset_files
SET "path" = REPLACE("path", 'upload/', '/share/immich/')
WHERE "path" LIKE 'upload/%';
```

I was updating the wrong column, in the wrong table.

![image](https://github.com/user-attachments/assets/d0657286-ccb8-406c-8735-aed70c4f1d17)

so that also explains why the symlink was just in the folder called uploads/ as the old path was relative.
